[Docs][EPP] Doc for request handling and control (#1128)
llm-d/llm-d: docs/wip-docs-new/architecture/core/epp/request-handling.md

        %% 3. Request Handling
        subgraph SubHandling ["RequestHandling Control"]
            direction TB
            subgraph SubPre ["Pre-Schedule Handling"]
                Prep[DataProducer.PrepareRequestData]:::statemgmt
                Admit[Admitter.AdmitRequest]:::statemgmt
            end

            subgraph SubPostSched ["Post-Scheduling Handling"]
                PreReq[PreRequest.PreRequest]:::postsched
                RespHead[ResponseHeaderProcessor.ResponseHeader]:::postsched
                RespBody[ResponseBodyProcessor.ResponseBody]:::postsched
            end
        end

        %% 4. Scheduler
        Sched[Scheduler.Schedule]:::scheduler    
    end
    Forward[Selected endpoints]

    %% Connections
    Req --> Parse
    Parse -->|"InferenceRequest"| FC
    
    %% Rejection Paths
    FC --> Prep
    FC -->|"Rejected/Evicted"| Reject
    
    Prep --> Admit
    Admit -->|"Admit"| Sched
    Admit -->|"Denied"| Reject
    
    %% Success Paths
    Sched -->|"SchedulingResult"| PreReq
    
    PreReq --> Forward
    Forward --> RespHead
    RespHead --> ParseResp
    ParseResp --> RespBody
    RespBody --> Client
```

> #### Core Components

Reviewer: As mentioned above I think we should focus on the components that the user cares about and will typically configure, which are the Parser, Admitter and DataProducer. I think the other extensions should only be documented in the framework developer guide (they are implementation details of those key components). This also makes it easier to align with the configuration guide.

Author: updated

> *   **Parser**: Responsible for parsing the request and response payloads to InferenceRequest, a structured internal representation of the incoming request.

Reviewer: ```suggestion
*   **Parser**: Responsible for parsing the incoming request to structured internal representation consumable by the scheduler, and parsing the response to extract usage data if reported by the model server.
```

* **Pre-Scheduling**: 
    * **DataProducer**: A pluggable extension that allows customizing request pre-processing and producing per-request state needed for scheduling, such as tokenization, prefix-cache matches, predicted processing latency etc..

>     * **Admitter**: Decides whether to admit a request based on criteria like latency SLOs. Runs after data production but before scheduling.

Reviewer: explain that request that don't pass the admitter will get rejected, while the ones that do proceed to the scheduling phase

Author: updated

* **Post-Scheduling**: 

>     * **PreRequest**: Executes after `SchedulingResult` is generated but before routing to the model server.

Reviewer: ```suggestion
    * **PreRequest**: Executes after `SchedulingResult` is generated but before passing the request back to the proxy to forward it to the model server.
```

    * **ResponseHeaderProcessor**: Triggered after response headers are successfully received. 
    * **ResponseBodyProcessor**: The primary interface for processing response data. It handles both streaming and non-streaming responses. For streaming responses, it processes each data chunk, with `EndOfStream` (EOS) set to true on the final chunk. For non-streaming responses, it runs exactly once with `EndOfStream` set to true.

---

### Concrete Plugins

#### Parsers
*   **[`openai-parser`](placeholder-link)**: The default parser supporting the OpenAI API. It parses request payloads to extract model name and prompts, and response payloads to extract usage data (tokens). It supports the following endpoints:
    *   `/conversations`
    *   `/responses`
    *   `/chat/completions`
    *   `/completions`
    *   `/embeddings`
*   **[`vllmgrpc-parser`](placeholder-link)**: A parser designed to handle requests specifically for the vLLM gRPC API. It supports:
    *   `Generate`
    *   `Embed`
*   **[`passthrough-parser`](placeholder-link)**: A model-agnostic parser that supports any request format by passing the request body through without interpretation.

> #### Request Control Plugins

Reviewer: I would remove this and remove one # from the two sections below

# EPP Request Handling and Control

The EPP Request Handling and Control component manages the lifecycle of an LLM request before and after the scheduling phase. It handles parsing the request payload, producing data for scheduling decisions, admission control, and processing the response from the model server.

### Architecture Overview

The request handling flow follows a structured sequence of extension points:

> ```mermaid

Reviewer: We should clarify how those are invoked with respect to flow control, they are kinda interleaved, right?

Author: updated, flow control should be right after Parser based on the logic

*   **PrepareDataPlugin**: Produces data needed for scheduling, such as latency predictions or cache state, before the scheduling decision is made.
*   **AdmissionPlugin**: Decides whether to admit a request based on criteria like latency SLOs.
*   **PreRequest**: Hook called after scheduling but before the request is forwarded to the selected endpoint.
*   **ResponseHeader**: Hook called by the director after response headers are successfully received, indicating the beginning of response handling by the model server.
*   **ResponseBody**: The primary hook for processing response data. It is called for every data chunk in a streaming response, or exactly once for non-streaming responses.

### Extension Points

Implemented via plugin interfaces in `pkg/epp/framework/interface`:

1.  **`Parser`**: Parses requests and responses. It extracts usage data from the response (e.g., for reporting or analysis). Supported by plugins in `pkg/epp/framework/plugins/requesthandling/parsers`.

> 2.  **`PrepareDataPlugin`**: Produces data for scheduling.

Reviewer: we need to rename this to DataProducer

Author: done

> 3.  **`AdmissionPlugin`**: Controls admission of requests.

Reviewer: we should rename to Admitter

Author: done

> 4.  **`PreRequest`**: Runs before forwarding.

Reviewer: We need to find a better name for this plugin, PreRequest is very confusing.

Author: currently I kept as is. Will update after the EPP code is updated and get a better name

5.  **`ResponseHeader`**: Runs on receiving headers.

> 6.  **`ResponseBody`**: Runs on receiving body chunks.

Reviewer: we should suffix them with Processor, `ResponseHeaderProcessor` and `ResponseBodyProcessor`

Author: done

#### Request Control Plugins
Located in `pkg/epp/framework/plugins/requestcontrol`.

##### Admission Plugins
*   **[`latency-slo-admitter`](placeholder-link)**: Rejects sheddable requests (priority < 0) when no endpoint can meet latency SLO constraints. It admits if at least one endpoint has valid predictions meeting SLOs, is idle, or is "cold" (<2% KV cache). Non-sheddable requests always bypass admission.

##### Data Producers
*   **[`predicted-latency-producer`](placeholder-link)**: Trains XGBoost models via a sidecar and generates per-endpoint TTFT/TPOT predictions. It calculates SLO headroom, collects training data, and tracks per-endpoint running request queues.
*   **[`inflight-load-producer`](placeholder-link)**: Tracks the number of in-flight requests and estimated tokens for each endpoint. It increments counts in `PreRequest` and decrements them in `ResponseBody` on end-of-stream.
*   **[`approx-prefix-cache-producer`](placeholder-link)**: Prepares data for approximate prefix cache aware scheduling by hashing prompts in blocks and matching them against an indexer of cached prefixes on servers.

> ##### Reporters

Reviewer: there is no "reporter" plugins, I would just remove this for now.

Author: Removed

# EPP Request Handling and Control

> The EPP Request Handling and Control component manages the lifecycle of an inference request before and after the scheduling phase. It handles parsing the request payload, producing data for scheduling decisions, admission control, and processing the response from the model server.

Reviewer 2: following the other docs, we should have a section called

```
## Functionality

>> explains ** what ** the component does

## Design

>> explains ** how ** the component is implemented
```

Author: Added the sections

### Architecture Overview

The request handling flow follows a structured sequence of extension points:

```mermaid
flowchart TD
    Req[Incoming Request] --> Parse[Parser.ParseRequest]
    Parse -->|InferenceRequest| FC[FlowControl.EnqueueAndWait]
    FC --> Prep[DataProducer.PrepareRequestData]
    FC -->|Rejected/Evicted| Reject[Reject Request]
    Prep --> Admit[Admitter.AdmitRequest]
    Admit --> S[Scheduler.Schedule]
    Admit -->|Denied| Reject
    S -->|SchedulingResult| PreReq[PreRequest.PreRequest]
    PreReq --> Route[Route to Model Server]
    Route --> RespHead[ResponseHeaderProcessor.ResponseHeader]
    RespHead --> ParseResp[Parser.ParseResponse]
    ParseResp --> RespBody[ResponseBodyProcessor.ResponseBody]
    RespBody --> Client[Return to Client]

> ```

Reviewer: give the request handling and control components a different color pls

Author: restructured graph, please take a look

Reviewer: Thanks a lot! Looks cluttered unfortunately and not aesthetically appealing. I think the problem is that we are trying to include every extension point, but not all of this necessary since not all of them a user cares about; I think we should only show the Parser, Admitter and DataProducer.

Author: Yes, I removed other parts and the graph is simpler, I didn't put flowControl out of the requestHandling Box because it will make the graph complex. PTAL

> #### Core Components

Reviewer: keep only the components related to request handling and control.

Author: done

> *   **Parser**: Responsible for parsing the request and response payloads to internal InferenceRequest.

Reviewer: ```suggestion
*   **Parser**: Responsible for parsing the request and response payloads to InferenceRequest, a structured internal representation of the incoming request.
```

> *   **FlowControl**: The main gatekeeper that calls `EnqueueAndWait` to queue requests and wait for capacity, enforcing priority and fairness.

Reviewer: remove pls

> *   **DataProducer**: Produces data needed for scheduling decision.

Reviewer: ```suggestion
*   **DataProducer**: A pluggable extension that allows customizing request pre-processing and producing per-request state needed for scheduling, such as tokenization, prefix-cache matches, predicted processing latency etc..
```

> *   **Admitter**: Decides whether to admit a request based on criteria like latency SLOs. Runs after data production but before scheduling.

Reviewer: ```suggestion
*   **Admitter**: Decides whether to admit a request based on criteria like latency SLOs. Runs after the data producer plugins but before request scheduling.
```

> *   **Scheduler**: Assigns the request to target endpoints.

Reviewer: remove pls

*   **PreRequest**: Hook called after `SchedulingResult` is generated but before routing to the model server.

> *   **ResponseHeaderProcessor**: Hook called after response headers are successfully received.

Reviewer: ```suggestion
*   **ResponseHeaderProcessor**: A hook called after response headers are successfully received. 
```

> *   **ResponseBodyProcessor**: The primary hook for processing response data. It handles both streaming and non-streaming responses: for streaming responses, it is called for each data chunk, with `EndOfStream` (EOS) set to true on the final chunk; for non-streaming responses, it is called exactly once with `EndOfStream` set to true.

Reviewer: ```suggestion
*   **ResponseBodyProcessor**: The primary hook for processing response data. It handles both streaming and non-streaming responses. For streaming responses, it is called for each data chunk, with `EndOfStream` (EOS) set to true on the final chunk. For non-streaming responses, it is called exactly once with `EndOfStream` set to true.
```

> ### Extension Points

Reviewer: This is a repeat of the above section, we should keep either this or the section above.

> The EPP framework provides extension points grouped into different packages under `pkg/epp/framework/interface`:

Reviewer: remove this pls

#### Request Handling (`requesthandling`)

*   **`Parser`**: Responsible for interpreting request and response payloads. Plugins implement this to support different API protocols (e.g., OpenAI, vLLM gRPC).

> #### Flow Control Layer

Reviewer: remove this section pls, not related to request handling per se

*   **Pluggable Policies**: Supports pluggable **Fairness Policies** (e.g., Round Robin), **Ordering Policies** (e.g., FCFS), and **Saturation Detectors**.
*   **More Details**: See the [Flow Control Guide](placeholder-link) for a complete overview.

#### Request Control (`requestcontrol`)

These plugins allow customizing the request lifecycle and scheduling decisions:

*   **`DataProducer`**: Executes before scheduling to gather or predict data (e.g., latency, cache state) needed by the scheduler.
*   **`Admitter`**: Makes admission decisions based on current pool state and request metadata (e.g., rejecting sheddable requests under load).
*   **`PreRequest`**: Runs after a scheduling decision is made but before the request is forwarded to the selected endpoint.
*   **`ResponseHeaderProcessor`**: Invoked when response headers are received from the model server.
*   **`ResponseBodyProcessor`**: Invoked for each chunk of the response body (or once for non-streaming), allowing processing and extraction of usage metrics.

---

### Concrete Plugins

#### Parsers
Located in `pkg/epp/framework/plugins/requesthandling/parsers`.

> *   **[`openai-parser`](placeholder-link)**: The default parser supporting the OpenAI API. It parses request payloads to extract model name and prompts, and response payloads to extract usage data (tokens).

Reviewer 2: which specific endpoints?

Author: updated
